### PR TITLE
Allow node scripts to return promises

### LINKED ISSUE #158: Allow returning promises from functions as opposed to context.done()
A more modern, industry accepted pattern for handling async operations is to use promises. Functions should allow developers to return a promise, rather than having to call context.done().

### PR DESCRIPTION
**Pull request by mamaso**
resolves #158

### CODE REVIEW

File: src/WebJobs.Script/Description/Node/functions/functions.js [hunk 1]
```diff
@@ -65,7 +69,12 @@ function createFunction(f) {
         inputs.unshift(context);
         delete context._inputs;
 
-        f.apply(null, inputs);
+        var userReturn = f.apply(null, inputs);
```
Comment (mathewc) [inline src/WebJobs.Script/Description/Node/functions/functions.js]: nit: `var result =`

File: src/WebJobs.Script/Description/Node/functions/functions.js [hunk 2]
```diff
@@ -65,7 +69,12 @@ function createFunction(f) {
...
+        if (userReturn && typeof userReturn.then === 'function') {
```
Comment (mathewc) [inline src/WebJobs.Script/Description/Node/functions/functions.js]: nit: `util.isFunction(userReturn.then)`

File: src/WebJobs.Script/Description/Node/functions/functions.js [hunk 3]
```diff
@@ -65,7 +69,12 @@ function createFunction(f) {
...
+            context._promise = true;
+            userReturn.then((result) => context.done(null, result))
```
Comment (mathewc) [inline src/WebJobs.Script/Description/Node/functions/functions.js]: Our `context.done` programming model allows a result to be returned, which is an object whose properties map to binding names (e.g. `context.done(null, { output: "FooBar"}`). It appears you're allowing a promise to conform to this and also return a result object, but I want to be sure :)

File: test/WebJobs.Script.Tests/TestScripts/Node/Scenarios/index.js [hunk 4]
```diff
@@ -18,6 +18,16 @@ var assert = require('assert');
     else if (scenario == 'promiseResolve') {
         Promise.resolve().then(() => context.done());
     }
+    else if (scenario == 'promiseApiResolves') {
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/TestScripts/Node/Scenarios/index.js]: You see how we've been limping along with full E2E tests for our Node interop shim. +1000 extra credit points if you can put in place some Mocha tests for this stuff that run automatically on the CI machine :)

**Comment by mathewc** [comment]
It'd be great if we either modified one of our existing [samples](https://github.com/Azure/azure-webjobs-sdk-script/tree/dev/sample) or added a new sample that shows of this new capability. We often point users to these samples to demonstrate things, so I'd like to be able to point someone to a sample that shows promises.

File: src/WebJobs.Script/Description/Node/functions/functions.js [hunk 5]
```diff
@@ -39,7 +39,11 @@ function createFunction(f) {
 
         context.done = function (err, result) {
             if (context._done) {
-                context.log("Error: 'done' has already been called. Please check your script for extraneous calls to 'done'.");
+                if (context._promise) {
+                    context.log("Error: Choose either to return a promise or call 'context.done'.  Do not use both in your script.");
+                } else {
+                    context.log("Error: 'done' has already been called. Please check your script for extraneous calls to 'done'.");
```
Comment (mathewc) [inline src/WebJobs.Script/Description/Node/functions/functions.js]: Let's make these error messages consistent by changing this one to reference 'context.done' as well. However, context is actually a variable name that users can change, so perhaps both error messages should just say 'done'?

File: test/WebJobs.Script.Tests/NodeEndToEndTests.cs [hunk 6]
```diff
@@ -733,6 +733,72 @@ public void ExcludedFunction_NotAddedToHost()
...
+                { "scenario", "promiseApiRejects" }
+            };
+
+            Task t = Fixture.Host.CallAsync("Scenarios",
+                new Dictionary<string, object>()
+                {
+                    { "input", input.ToString() }
+                });
+
+            Task result = await Task.WhenAny(t, Task.Delay(5000));
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/NodeEndToEndTests.cs]: Why do we need a 5 second delay here?

File: src/WebJobs.Script/Description/Node/functions/functions.js (hunk 2 above)
Comment (mamaso) [inline src/WebJobs.Script/Description/Node/functions/functions.js]: util.isFunction is deprecated, so I think the typeof check is a better choice.

As for promises and 'then': http://stackoverflow.com/a/27746324

File: test/WebJobs.Script.Tests/TestScripts/Node/Scenarios/index.js (hunk 4 above)
Comment (mamaso) [inline test/WebJobs.Script.Tests/TestScripts/Node/Scenarios/index.js]: Yep!  What do I need to get admin access for appveyor?

File: test/WebJobs.Script.Tests/NodeEndToEndTests.cs (hunk 6 above)
Comment (mamaso) [inline test/WebJobs.Script.Tests/NodeEndToEndTests.cs]: In a nutshell, the nextTick issue.  If nextTick or Promises start failing the tests will hang, so we need the timeout to shortcircuit & fail the test.

File: src/WebJobs.Script/Description/Node/functions/functions.js (hunk 3 above)
Comment (mamaso) [inline src/WebJobs.Script/Description/Node/functions/functions.js]: It should work with the implementation, I'll add in a test for it

**Comment by mamaso** [comment]
@mathewc I'll add a sample when this is merged / we release.

File: test/WebJobs.Script.Tests/TestScripts/Node/Scenarios/index.js (hunk 4 above)
Comment (fabiocav) [inline test/WebJobs.Script.Tests/TestScripts/Node/Scenarios/index.js]: I can work on giving you the access you need there.

File: test/WebJobs.Script.Tests/NodeEndToEndTests.cs [hunk 7]
```diff
@@ -734,6 +734,72 @@ public void ExcludedFunction_NotAddedToHost()
...
+            Assert.Same(t, result);
+            Assert.Equal(true, t.IsFaulted);
+            Assert.Equal("reject", t.Exception.InnerException.InnerException.Message);
```
Comment (fabiocav) [inline test/WebJobs.Script.Tests/NodeEndToEndTests.cs]: Is this dealing with nested `AggregateExceptions` here? If so, just use `Flatten` to clean this up a bit.

Comment (mamaso) [inline test/WebJobs.Script.Tests/NodeEndToEndTests.cs]: Unfortunately `Flatten` & `GetBaseException` both fail in this case, so I'll leave as is

File: src/WebJobs.Script/Description/Node/functions/functions.js (hunk 3 above)
Comment (mamaso) [inline src/WebJobs.Script/Description/Node/functions/functions.js]: @mathewc Added the bindings test, see: https://github.com/Azure/azure-webjobs-sdk-script/pull/673/files#diff-c79b5147f12f44c2f0ca539cbdb3ac21R6

File: test/WebJobs.Script.Tests/TestScripts/Node/Scenarios/index.js (hunk 4 above)
Comment (mamaso) [inline test/WebJobs.Script.Tests/TestScripts/Node/Scenarios/index.js]: Thanks Fabio! I'll get in the mocha testing in a separate PR

**Pull request merged**